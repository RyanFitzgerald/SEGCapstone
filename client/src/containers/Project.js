import React from 'react';
import { Switch, Route } from 'react-router-dom';
import * as api from '../api';
import arraySort from 'array-sort';

// Import client components
import Submenu from '../components/projects/Submenu';
import Add from '../components/projects/Add';
import Directory from '../components/projects/Directory';
import View from '../components/projects/View';
import Edit from '../components/projects/Edit';
import Note from '../components/projects/Note';
import Product from '../components/projects/Product';
import CostUpdate from '../components/projects/CostUpdate';
import Photo from '../components/projects/Photo';
import File from '../components/projects/File';

class Project extends React.Component {
  constructor() {
    super();

    // Bind functions
    this.setActiveSubtab = this.setActiveSubtab.bind(this);
    this.getTypes = this.getTypes.bind(this);
    this.getClients = this.getClients.bind(this);
    this.getProjects = this.getProjects.bind(this);
    this.removeFromProjects = this.removeFromProjects.bind(this);
    this.contentLoaded = this.contentLoaded.bind(this);
    this.renderError = this.renderError.bind(this);
    this.sortByKey = this.sortByKey.bind(this);

    // State
    this.state = {
      activeSubtab: 1,
      types: null,
      clients: null,
      projects: null,
      sort: {
        name: null,
        client: null,
        status: null
      }
    };
  }

  componentDidMount() {
    // Set Page Title
    document.title = 'Project Overview | Renovaction';

    // Update tab
    this.props.setActiveTab(2);

    // Get project types
    this.getTypes();

    // Get clients
    this.getClients();

    // Get projects
    this.getProjects(false);
  }

  setActiveSubtab(tab) {
    this.setState({activeSubtab: tab});
  }

  getTypes() {
    api.getTypes().then(types => {
      this.setState({ types });
    });
  }

  getClients() {
    api.getClients(false).then(result => {
      this.setState({ clients: result });
    });
  }

  getProjects(query) {
    api.getProjects(query).then(result => {
      const projects = result;
      this.setState({ projects });
    });
  }

  removeFromProjects(id) {
    const projects = [...this.state.projects];
    const updated = projects.filter(el => {
      return el._id !== id;
    });
    this.setState({ projects: updated });
  }

  sortByKey(array, key) {
    let asc = 'asc';
    let desc = 'desc'
    let sortOrder = {name: null, client: null, status: null};
    const arr = Object.keys(array).map((k) => array[k]);
    var sortedArray = [];

    if (key === 'name') {
      if (this.state.sort.name === asc) {
        sortOrder.name = desc;
        sortedArray = arraySort(arr, key, {reverse: true});
      } else {
        sortOrder.name = asc;
        sortedArray = arraySort(arr, key);
      }
    } else if (key === 'client.name') {
      if (this.state.sort.client === asc) {
        sortOrder.client = desc;
        sortedArray = arraySort(arr, key, {reverse: true});
      } else {
        sortOrder.client = asc;
        sortedArray = arraySort(arr, key);;
      }
    } else if (key === 'status') {
      if (this.state.sort.status === asc) {
        sortOrder.status = desc;
        sortedArray = arraySort(arr, key, {reverse: true});
      } else {
        sortOrder.status = asc;
        sortedArray = arraySort(arr, key);;
      }
    }

    // Update state
    this.setState({projects: sortedArray, sort: sortOrder});
  }

  renderError(formError) {
    if (!formError) return;
    return (
      <div className="flash flash--warn">
        <p>{formError}</p>
      </div>
    );
  }

  contentLoaded() {
    return this.state.clients && this.state.projects && this.state.types;
  }

  render() {
    return (
      <div>
        <Submenu activeSubtab={this.state.activeSubtab} level={this.props.level} checkLevel={this.props.checkLevel}/>

        <Switch>
          <Route exact path="/projects" render={() =>
            <Directory setActiveSubtab={this.setActiveSubtab} projects={this.state.projects} types={this.state.types} getProjects={this.getProjects} sortByKey={this.sortByKey} />
          }/>
          <Route path="/projects/add" render={(location) =>
            <Add setActiveSubtab={this.setActiveSubtab} types={this.state.types} clients={this.state.clients} addNotification={this.props.addNotification} renderError={this.renderError} location={location} getProjects={this.getProjects} level={this.props.level} checkLevel={this.props.checkLevel}/>
          }/>
          <Route path="/projects/:id/edit" render={(location) =>
            <Edit setActiveSubtab={this.setActiveSubtab} types={this.state.types} clients={this.state.clients} addNotification={this.props.addNotification} renderError={this.renderError} location={location} getProjects={this.getProjects}/>
          }/>
          <Route path="/projects/:id/note" render={(location) =>
            <Note setActiveSubtab={this.setActiveSubtab} addNotification={this.props.addNotification} renderError={this.renderError} location={location} />
          }/>
          <Route path="/projects/:id/product" render={(location) =>
            <Product setActiveSubtab={this.setActiveSubtab} addNotification={this.props.addNotification} renderError={this.renderError} location={location} />
          }/>
          <Route path="/projects/:id/photo" render={(location) =>
            <Photo setActiveSubtab={this.setActiveSubtab} addNotification={this.props.addNotification} renderError={this.renderError} location={location} />
          }/>
          <Route path="/projects/:id/file" render={(location) =>
            <File setActiveSubtab={this.setActiveSubtab} addNotification={this.props.addNotification} renderError={this.renderError} location={location} />
          }/>
          <Route path="/projects/:id/update" render={(location) =>
            <CostUpdate setActiveSubtab={this.setActiveSubtab} addNotification={this.props.addNotification} renderError={this.renderError} location={location} />
          }/>
          <Route path="/projects/:id" render={(location) =>
            <View setActiveSubtab={this.setActiveSubtab} addNotification={this.props.addNotification} location={location} removeFromProjects={this.removeFromProjects}/>
          }/>
        </Switch>
      </div>
    );
  }
}

export default Project;