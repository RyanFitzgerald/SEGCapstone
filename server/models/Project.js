/*
  Project Model
  - This file contains the model for projects
*/

const mongoose = require('mongoose');
mongoose.Promise = global.Promise;
const mongodbErrorHandler = require('mongoose-mongodb-errors');
require('dotenv').config({ path: 'variables.env' });
const NodeGeocoder  = require('node-geocoder');
var options = {
  provider: 'google',
  apiKey: process.env.GEOCODE_KEY,
  formatter: null
};
var geocoder = NodeGeocoder(options);

const projectSchema = new mongoose.Schema({
  created: {
    type: Date,
    default: Date.now
  },
  name: {
    type: String,
    unique: true,
    required: 'A project name must be provided!',
    trim: true
  },
  street: {
    type: String,
    required: 'A street must be provided!',
    trim: true
  },
  postalCode: {
    type: String,
    required: 'A postal code must be provided!',
    trim: true
  },
  city: {
    type: String,
    required: 'A city must be provided!',
    trim: true
  },
  soldDate: {
    type: Date,
    required: 'A sold date must be provided'
  },
  startDate: Date,
  endDate: Date,
  cashinDate: Date,
  labourCost: Number,
  materialsCost: Number,
  actualCost: Number,
  status: String,
  type: [{
    type: mongoose.Schema.ObjectId,
    ref: 'Type',
    required: 'A type must be provided'
  }],
  subtype: [{
    type: mongoose.Schema.ObjectId,
    ref: 'SubType'
  }],
  client: {
    type: mongoose.Schema.ObjectId,
    ref: 'Client',
    required: 'A client must be provided'
  },
  location: {
    type: {
      type: String,
      default: 'Point'
    },
    coordinates: {
      type: [Number],
      default: [-75.69719309999999, 45.4215296]
    },
    address: {
      type:String,
      default: 'Ottawa, ON, Canada'
    }
  }
  /*createdBy: {
    type: mongoose.Schema.ObjectId,
    ref: 'User',
    required: 'A user must be provided'
  }*/
}, {
  toJSON: { virtuals: true }, // Show virtuals explicitly, but not needed
  toObject: { virtuals: true } // Show virtuals explicitly, but not needed
});

projectSchema.plugin(mongodbErrorHandler);

projectSchema.virtual('notes', {
  ref: 'ProjectNote',
  localField: '_id', // Which field on project
  foreignField: 'project' // which field on note
});

projectSchema.virtual('products', {
  ref: 'Product',
  localField: '_id', // Which field on project
  foreignField: 'project' // which field on product
});

projectSchema.virtual('updates', {
  ref: 'CostUpdate',
  localField: '_id', // Which field on project
  foreignField: 'project' // which field on update
});


projectSchema.pre('save', async function(next) {
  const street = this.street;
  const postalCode = this.postalCode;

  geocoder.geocode({address: street, country: 'Canada', zipcode: postalCode})
    .then(res => {
      let long = res[0].longitude;
      let lat = res[0].latitude;
      let address = res[0].formattedAddress;

      this.location.coordinates = [long, lat];
      this.location.address = address;

      next();
    })
    .catch(err => {
      console.error(err);
      next();
    });  
});

module.exports = mongoose.model('Project', projectSchema);