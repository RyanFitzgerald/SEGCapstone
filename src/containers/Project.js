import React, { Component, PropTypes } from 'react';
import {withRouter, Route, Switch, Link} from 'react-router-dom';
import * as api from '../api';

// Import Client Components
import Summary from '../components/projects/Summary';
import Add from '../components/projects/Add';
import Edit from '../components/projects/Edit';
import View from '../components/projects/View';
import Search from '../components/projects/Search';

class Project extends Component {

    constructor(props) {
        super(props);

        // States
        this.state = {
            headerTab: 1,
            cities: null,
            clients: null,
            projects: null,
            projectTypes: null
        };
    }

    setHeaderTab = (tab) => {
        this.setState({
            headerTab: tab
        });
    }

    componentDidMount() {
        // Update page title
        document.title = 'Projects Overview';

        // Update Active Tab
        this.props.setActiveTab(2);

        // Update Header Title
        this.props.setHeaderTitle('Projects');

        // Get cities
        this.getCities();

        // Get clients
        this.getClients();

        // Get projects
        this.getProjects();

        // Get project types
        this.getProjectTypes();
    }

    addProject = project => {
        api.addProject(project).then(resp => {
            this.getProjects();
            this.props.history.push(`/projects/${resp.projectID}`);
        });
    }

    editProject = project => {
        api.editProject(project).then(resp => {
            this.getProjects();
            this.props.history.push(`/projects/${resp.projectID}`);
        });
    }

    deleteProject = projectID => {
        api.deleteProject(projectID).then(() => {
            this.getProjects();
            this.props.history.push('/projects');
        });
    }

    getProjects = () => {
        api.getProjects().then(projects => {
            this.setState({
                projects
            });
        });
    }

    getCities = () => {
        api.getCities().then(cities => {
            this.setState({
                cities
            });
        });
    }

    getClients = () => {
        api.getClients().then(clients => {
            this.setState({
                clients
            });
        });
    }

    getProjectTypes = () => {
        api.getProjectTypes().then(projectTypes => {
            this.setState({
                projectTypes
            });
        });
    }

    searchProjects = projectName => {
        api.searchProjects(projectName).then(projects => {
            this.setState({
                projects
            });
        });
    };

    render() {
        return (
            <div className="projects">
                <div id="page-navigation">
                    <ul>
                        <li className={(this.state.headerTab == 1) ? 'active' : ''}>
                            <Link to="/projects">Overview</Link>
                        </li>
                        <li className={(this.state.headerTab == 2) ? 'active' : ''}>
                            <Link to="/projects/search">Search</Link>
                        </li>
                        <li className={(this.state.headerTab == 3) ? 'active' : ''}>
                            <Link to="/projects/add">Add</Link>
                        </li>
                    </ul>
                </div>
                {/* <!-- End page-navigation --> */}

                <Switch>
                    <Route exact path="/projects" render={() =>
                        <Summary setHeaderTab={this.setHeaderTab} />
                    } />
                    <Route path="/projects/search" render={() =>
                        <Search setHeaderTab={this.setHeaderTab} searchProjects={this.searchProjects} projects={this.state.projects} cities={this.state.cities} projectTypes={this.state.projectTypes} />
                    } />
                    <Route path="/projects/add" render={() =>
                        <Add setHeaderTab={this.setHeaderTab} addProject={this.addProject} cities={this.state.cities} clients={this.state.clients} projectTypes={this.state.projectTypes} />
                    } />
                    <Route path="/projects/:projectID/edit" render={({match}) =>
                        <Edit setHeaderTab={this.setHeaderTab} match={match} editProject={this.editProject} cities={this.state.cities} clients={this.state.clients} projectTypes={this.state.projectTypes} />
                    } />
                    <Route path="/projects/:projectID" render={({match}) =>
                        <View setHeaderTab={this.setHeaderTab} match={match} deleteProject={this.deleteProject} />
                    } />
                </Switch>
            </div>
        );
    }
}

Project.propTypes = {
    setActiveTab: PropTypes.func.isRequired,
    setHeaderTitle: PropTypes.func.isRequired,
    history: PropTypes.any
};

export default withRouter(Project);
