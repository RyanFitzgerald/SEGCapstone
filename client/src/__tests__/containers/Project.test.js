import React from 'react';
import Project from '../../Containers/Project';

describe('<Project />', () => {
  beforeAll(() => {
    // Add necessary info to user
    const user = {
      role: {
        level: 1
      }
    };
    localStorage.setItem('user', JSON.stringify(user));
  });

  it('should render correctly', () => {
    // Return empty objects on requests
    mock.onAny().reply(200, {});

    // Create mock functions / spies
    const setActiveTabMock = jest.fn();
    const getTypesSpy = jest.spyOn(Project.prototype, 'getTypes');
    const getClientsSpy = jest.spyOn(Project.prototype, 'getClients');
    const getProjectsSpy = jest.spyOn(Project.prototype, 'getProjects');

    // Shallow render it
    const wrapper = shallow(
      <Project setActiveTab={setActiveTabMock}/>
    );

    // Compare to old snapshot
    expect(shallowToJson(wrapper)).toMatchSnapshot();

    // Check set active tab is called and with correct value
    expect(setActiveTabMock).toHaveBeenCalledTimes(1);
    expect(setActiveTabMock.mock.calls[0][0]).toBe(2);

    // Check if component mounting methods were called
    expect(getTypesSpy).toHaveBeenCalled();
    expect(getClientsSpy).toHaveBeenCalled();
    expect(getProjectsSpy).toHaveBeenCalled();
  });
});

