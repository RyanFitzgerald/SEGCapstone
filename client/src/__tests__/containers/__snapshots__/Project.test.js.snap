// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Project /> should render correctly 1`] = `
<div>
  <Submenu
    activeSubtab={1}
  />
  <Switch>
    <Route
      exact={true}
      path="/projects"
      render={[Function]}
    />
    <Route
      path="/projects/add"
      render={[Function]}
    />
    <Route
      path="/projects/:id/edit"
      render={[Function]}
    />
    <Route
      path="/projects/:id/note"
      render={[Function]}
    />
    <Route
      path="/projects/:id/product"
      render={[Function]}
    />
    <Route
      path="/projects/:id/photo"
      render={[Function]}
    />
    <Route
      path="/projects/:id/file"
      render={[Function]}
    />
    <Route
      path="/projects/:id/update"
      render={[Function]}
    />
    <Route
      path="/projects/:id"
      render={[Function]}
    />
  </Switch>
</div>
`;
